girder/utility/filesystem_assetstore_adapter.py

```diff
@@ -320,31 +320,59 @@ def importFile(self, item, path, user, name=None, mimeType=None, **kwargs):
         file['imported'] = True
         return self.model('file').save(file)
 
-    def importData(self, parent, parentType, params, progress, user):
+    def _importDataAsItem(self, name, user, folder, path, files,
+                          reuseExisting=True):
+        item = self.model('item').createItem(
+            name=name, creator=user, folder=folder,
+            reuseExisting=reuseExisting)
+        for fname in files:
+            self.importFile(item, os.path.join(path, fname),
+                            user, name=fname)
+
+    def importData(self, parent, parentType, params, progress, user,
+                   leafFoldersAsItems):
         importPath = params['importPath']
 
         if not os.path.isdir(importPath):
             raise ValidationException('No such directory: %s.' % importPath)
 
-        for name in os.listdir(importPath):
+        def hasOnlyFiles(path, files):
+            return all(map(os.path.isfile, (os.path.join(path, fname)
+                                            for fname in files)))
+
+        listDir = os.listdir(importPath)
+        if leafFoldersAsItems and hasOnlyFiles(importPath, listDir):
+            self._importDataAsItem(os.path.basename(importPath.rstrip('/')),
+                                   user, parent, importPath, listDir)
+            return
+
+        for name in listDir:
             progress.update(message=name)
             path = os.path.join(importPath, name)
 
             if os.path.isdir(path):
-                folder = self.model('folder').createFolder(
-                    parent=parent, name=name, parentType=parentType,
-                    creator=user, reuseExisting=True)
-                self.importData(folder, 'folder', params={
-                    'importPath': os.path.join(importPath, name)
-                }, progress=progress, user=user)
+                localListDir = os.listdir(path)
+                if leafFoldersAsItems and hasOnlyFiles(path, localListDir):
+                    self._importDataAsItem(name, user, parent, path,
+                                           localListDir)
+                else:
+                    folder = self.model('folder').createFolder(
+                        parent=parent, name=name, parentType=parentType,
+                        creator=user, reuseExisting=True)
+                    self.importData(folder, 'folder', params={
+                        'importPath': os.path.join(importPath, name)},
+                        progress=progress, user=user,
+                        leafFoldersAsItems=leafFoldersAsItems)
             else:
                 if parentType != 'folder':
                     raise ValidationException(
-                        'Files cannot be imported directly underneath a %s.' %
+                        'Files cannot be imported directly'
+                        'underneath a %s.' %
```

Comment:
Ah I see, the lines were exactly 80 chars. Our style checker permits that, but it's probably better your way.